### ISSUE TITLE
Is there a way to defer validation until after __init__ and trigger it manually?

### ISSUE THREAD

**Issue by lsemel** [samuelcolvin/pydantic#559]
# Question

I'd like to defer validation until after `__init__`, ideally only when I call it manually.  This way pydantic models could be worked with similar to Django models in that they don't require all the required fields to be passed in when initializing them, i.e.

```
from typing import List
from pydantic import BaseModel, validate_model, ValidationError

class SomeModel(BaseModel):
    id: int
    title: str
    tags: List[str]

model = SomeModel()                                # this fails validation due to the missing required fields
model.id = 1
model.title = 'Title'
model.tags = ['foo', 'bar', 'baz']
try:
    validate_model(model, model.__values__)        # is there an instance method for this?
    print(model.json())
except ValidationError:
    pass

```

Right now, if I want to do something like this, I'd have to assemble the values in a `dict` and pass them to the constructor, which is ugly compared to the nice OO syntax pydantic provides.  For this trivial example that's not a big deal, but I need to assemble models with large sets of fields populated from a variety of functions.  This is currently done by passing around and updating lot of `dict`s, which is clunky and hard-to-read, and that's one of the reasons I want to move to the clearer syntax of dataclasses or pydantic in the first place.

One workaround is to give all the fields default values, but if they're ultimately required fields, then there's no way to actually validate their presence.

In other words, it'd be good if there were a way to validate a model when you need to, such as before serializing, but without requiring that it be valid at all moments throughout its existence.

**Comment by samuelcolvin** [comment]
Not impossible to do, but I doubt it would actually be used by many people.

Probably better to setup a your own custom dataclass which defers validation.

Simple example:

```py
from datetime import datetime
from pydantic.dataclasses import dataclass

@dataclass
class User:
    id: int
    name: str = 'John Doe'
    signup_ts: datetime = None

User.validate = User.__post_init__
User.__post_init__ = lambda _: None

user = User(id='xxx', signup_ts='2032-06-21T12:00')
print(user)
#> User(id='xxx', name='John Doe', signup_ts='2032-06-21T12:00')

user.validate()
#> ValidationError....
```

Let me know if that really doesn't work for you.

**Comment by dbichko** [comment]
For what it's worth, I would very much use that feature, for exactly the reasons @lsemel described.

**Comment by lsemel** [comment]
I ended up working around this by allowing all the fields to default to None.

**Comment by dmontagu** [comment]
@dbichko @lsemel Thanks to the change `__values__` -> `__dict__`, as of pydantic 0.32 (I believe), this is possible:

```python
from typing import Generic, TypeVar, TYPE_CHECKING, Dict, Any

from pydantic import BaseModel, ValidationError

T = TypeVar("T")


class DeferredValidationModel(BaseModel):
    if TYPE_CHECKING:
        __deferred_validation_kwargs__: Dict[str, Any]

    def __init__(self, **kwargs: Any) -> None:
        self.__dict__["__deferred_validation_kwargs__"] = kwargs

    def validate_after_deferral(self: T) -> T:
        super().__init__(**self.__deferred_validation_kwargs__)


class A(DeferredValidationModel):
    x: int

a = A(x="1")
print(a)
# A __deferred_validation_kwargs__={'x': '1'}

a.validate_after_deferral()
print(a)
# A x=1

a = A(x="a")
print(a)
# A __deferred_validation_kwargs__={'x': 'a'}

try:
    a.validate_after_deferral()
except ValidationError as exc:
    print(exc)
"""
1 validation error for A
x
  value is not a valid integer (type=type_error.integer)
"""
```
I couldn't figure out how to backport it to the older versions using `__values__` without running into recursion errors, but that might be possible.

**Comment by dmontagu** [comment]
(I edited my original answer with a typesafe version that works both in current pydantic and older pydantic.)

**Comment by ojomio** [comment]
Any plans to include the above snippet as a part of the library? Looks promising to me

**Comment by maxims94** [comment]
This feature would be great!

**Comment by kstohr** [comment]
This feature would lead to much cleaner code.

**Comment by samuelcolvin** [comment]
No, I have no plans to support this.

It **might** become easier to hack a solution to this in v2.

**Comment by pmbaumgartner** [comment]
I've had a need for this, just posting my solution here.

In my case, I have several steps of a data processing pipeline that create instances of my objects that are stored intermediately prior to validation, and I want to defer validating them to the end because if there's a `ValidationError` I want to write code to handle that error in one place (at the end of the pipeline) rather than handling it at several prior steps of my pipeline when the objects are created.

To clarify with the example below, I'm first creating objects with `instance = Range.construct(minimum=Fraction(3,1), maximum=Fraction(4,5))` at various points, then later calling `validated_instance = instance.reinit()` as I've implemented below.

All I have is this method on my model (ignore the model details in this example):

```
class Range(BaseModel):
    minimum: Optional[Fraction] = Field(...)
    maximum: Optional[Fraction] = Field(...)

    def reinit(self):
        self.__init__(**super().dict())
```

In my case, I was also overwriting `Range.dict()` (not shown) to serialize these objects correctly, so I needed to call `super()` to get the original `dict()` implementation. If you haven't overwritten the `dict()` method, you can implement the same thing by calling `self.dict()` rather than `super().dict()`.

**Comment by coolkau** [comment]
I constantly need a deferred validation when ingesting data from anything apart databases, which is 90% of the use-cases I have e.g.
- Load data from multiple jsons
- Extract data from text
- Data from multiple API calls
-  Messy data which is cleaned step by step beyond what makes sense to do it in one go using validators.

Data comes in one by one, you need an extra container to store the data, either a dict or a non-pydantic class containing the same fields. It is just double the work, hard to maintain. I have lots of projects containing two identical classes... 

It would be so nice to be able to say "dont to anything until I flip the switch".

Now that version 2 is here perhaps it is easier to implement?

**Comment by smarie** [comment]
Unfortunately [@dmontagu 's great proposal](https://github.com/pydantic/pydantic/issues/559#issuecomment-519686409) has two limitations:

- latest `mypy` (1.20) complains `error: "DeferredModel[Foo]" has no attribute "a"  [attr-defined]`
- deferred models can not easily be filled with more attributes between initial creation and final validation.

Here is an enhancement that seems to work:

```python
from typing import TypeVar, Dict, Any, Generic, Type

from pydantic import BaseModel, Field

T = TypeVar("T", bound=BaseModel)


class DeferredModel(Generic[T]):
    def __init__(self, type_: Type[T], kwargs: Dict[str, Any]):
        self.type_ = type_
        # self.kwargs = kwargs
        for n, v in kwargs.items():
            setattr(self, n, v)

    @property
    def kwargs(self) -> Dict[str, Any]:
        return {k: v for k, v in vars(self).items() if k != "type_"}

    def deferred_create(self) -> T:
        return self.type_(**self.kwargs)

    def __repr__(self):
        return f"{type(self).__name__}(type_={self.type_.__name__}, kwargs={self.kwargs})"


class DeferrableModel(BaseModel):
    @classmethod
    def defer(cls: Type[T], **kwargs: Any) -> T:
        """The output type is deliberately wrong so that mypy believes that the output is of type `T`"""
        return DeferredModel(cls, kwargs)  # type: ignore[return-value]

    def deferred_create(self: T) -> T:
        """The role of this function is for mypy to know it when a DeferredModel is returned as a T in `defer` above"""
        raise NotImplementedError("Please call `defer()` first to get an instance to fill with data, then call `deferred_create()` on the instance.")


class Foo(DeferrableModel):
    a: int
    b: str


# 0. Create an instance while not providing any contents
f = Foo.defer()
print(f)  # DeferredModel(type_=Foo, kwargs={})

f.a = 1
print(f)  # DeferredModel(type_=Foo, kwargs={'a': 1})

# 1. Create too soon - fields missing
try:
    f.deferred_create()
except Exception as e:
    print(e)  # 1 validation error for Foo   - b   Field required [type=missing, input_value={'a': 1}, input_type=dict]

f.b = "hello"
print(f)  # DeferredModel(type_=Foo, kwargs={'a': 1, 'b': 'hello'})

# 2. Nominal usage - Ready to create
g = f.deferred_create()  # --> Foo(a=1, b='hello')
print(repr(g))  # Foo(a=1, b='hello')

# 3. Create with unproper data in one field
f.b = 1
try:
    f.deferred_create()
except Exception as e:
    print(e)  # 1 validation error for Foo   - b   Input should be a valid string [type=string_type, input_value=1, input_type=int]

```

**Comment by smarie** [comment]
Oh and I tried this solution too from @samuelcolvin but it does not work with latest pydantic : https://github.com/pydantic/pydantic/issues/559#issuecomment-496538989
